pdns/cachecleaner.hh

```diff
@@ -147,7 +147,27 @@ template <typename T> uint64_t pruneLockedCollectionsVector(vector<T>& maps, uin
     }
     totErased += erased;
   }
+  
+  toTrim = toTrim - totErased;
+
+  if (toTrim > 0) {
+    //Just trowing away the old stuff wasn't sufficient, start from the top
+    
+    for(auto& mc : maps) {
+      WriteLock wl(&mc.d_mut);
+      auto& sidx = boost::multi_index::get<2>(mc.d_map);
+      uint64_t erased = 0, lookedAt = 0;
+      for(auto i = sidx.begin(); i != sidx.end(); lookedAt++) {
+        i = sidx.erase(i);
+        erased++;
 
+        if(erased > toTrim / maps.size())
+          break;
```

Comment:
One potential issue here is that this code will always nuke the content of the shards in the same order until it has nuked enough entries, which means the last shards will often not be reached. Perhaps we should start with a random shard every time?
It's not that bad with the existing code because we are expunging expired entries, but that would be an improvement there as well.

Reply:
I think it would even be better to simply have the new code start from the bottom and keep removing the expired entries from the top. This will cause no extra overhead at all, and the two combined we have a good chance having checked the complete list and a small chance of going over entries twice.

Reply:
Do you mean iterating over the shards in reverse order to remove all entries, during the second phase? That might help but I'm afraid it would still be very skewed.
And if you mean removing the entries inside the shards in reverse order it would remove the most recently used ones, which would not be a good idea.

Reply:
Sorry, I wasn't taking into account the fact that most recently used would be at the bottom.

I'm very new to this code, so sorry if I'm mistaking. But taking that order into account, I would think it would be beneficial to always start at the top. Randomly starting lower to the bottom would mean we're deleting more recently used entries?

Reply:
You are mixing two things, but that's because I wasn't clear enough about that, sorry. `maps` is a vector of shards, and I think it would make sense to start with a random shard for the final cleanup. `d_map` is a multi-indexed object whose second index should be a LRU. It's a LRU for the query and packet cache of the recursor, but surprisingly it looks like it was never a LRU in the auth.. It should be.

Reply:
Sorry if this is becoming a back-and-forth, but aren't we always going over all elements in maps, thus always going over all the shards?

Reply:
Oh, right, you remove the same number of entries from each shard so the order doesn't matter. Something in the initial version made me think that you could have removed a lot of entries from the first shards and none at all from the last ones but I was clearly mistaken, sorry about that.